README.md
# About SSHExtension

[![Join the chat at https://gitter.im/Vitalykondratiev/vscode-sshextension](https://badges.gitter.im/Vitalykondratiev/vscode-sshextension.svg)](https://gitter.im/Vitalykondratiev/vscode-sshextension?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)

[![Latest Release](https://vsmarketplacebadge.apphb.com/version/kondratiev.sshextension.svg)](https://marketplace.visualstudio.com/items?itemName=kondratiev.sshextension)
[![Installs](https://vsmarketplacebadge.apphb.com/installs-short/kondratiev.sshextension.svg)](https://marketplace.visualstudio.com/items?itemName=kondratiev.sshextension)

This extension allows you to open an SSH connection in the integrated terminal.
The extension was created in order to have access to the SSH in conjunction with the already available access to the FTP.  

## Features

You can use configurations from another extensions (e.g. ftp-simple)  
The connection opens in a new instance of the integrated terminal.  
SSH port forwarding.

## How to use

### Open terminal from server list
- Open the Command Palette (usually `F1` or `Ctrl+Shift+P`).  
- Select the command `SSHExtension: Open SSH Connection`.  
- Select a server from the list.

![Demo Open connection from list](./images/open_connection_from_list.gif)

### Fast open terminal
- Open workspace with project mapped to server
- Open any project file or go to already opened editor tab  
- Click on "Open SSH on \<servername>" button

![Demo Open connection from list](./images/open_fast_connection.gif)

### SSH port forwarding
- Open the Command Palette (usually `F1` or `Ctrl+Shift+P`).  
- Select the command `SSHExtension: SSH Port Forwarding`.  
- Select a forwarding type from: `Local to remote` (-L), `Local to remote` (-R), `SOCKS` (-D), `Recently used` (if exists saved arguments).  
- Enter the required parameters on request.  
- (Optionally) You can save your selections for faster port forwarding in the future.

![Demo Open connection from list](./images/port_forwarding.gif)

To add a server, see Settings section.

## Requirements
  
You should still have an ssh agent, not necessarily that it is available in the entire system. it is important that it is accessible from the integrated VSCode terminal.

## Settings (for servers)

You can use ready-made config file from this extensions (if you use):
- ftp-simple ([see info about configuring](https://marketplace.visualstudio.com/items?itemName=humy2833.ftp-simple#user-content-config-setting-example), servers with `"type": "sftp"` only).

Or you can use extension settings simply add `sshextension.serverList` directive.

## Extension settings

#### sshextension.serverList

  * Type: `Array`
  * Defaut: `[]`

You can describe servers config in this parameter as array of objects.  
Server object parameters:  
- **name** _(string)_* - name of server (showing in picks list if `showHostsInPickLists` is `false`).  
- **host** _(string)_* - server hostname.
- **port** _(number)_ - SSH port. 
- **username** _(string)_* - username for authentication.
- **password** _(string)_ - password for authentication.
- **privateKey** _(string)_ - string that contains a path to private key.
- **project**  _(object)_ - specify local workspace path and server root path for fast terminal open.
- **path** _(string)_ - used for change directory after server connection.
- **customCommands** _(array of strings)_ - specifies custom commands which will execute on session start

For example:
```json
{
    "sshextension.serverList": [
        {
            "name": "Example server",
            "host": "example.com",
            "port": 22,
            "username": "user",
            "privateKey": "D:\\id_rsa",
            "project": {
                "D:/projects/project": "/home/user/project",
                "D:/projects/yet_another_project": "/home/user/yet_another_project"
            },
            "path": "/",
            "customCommands": [
              "pwd"
            ]
        },
        ...
    ]
}
```

#### sshextension.customCommands

  * Type: `Array`
  * Defaut: `[]`

Specifies custom commands which will execute on session start.  
For example:
```json
{
  "sshextension.customCommands": [
    "pwd",
    "ls"
  ]
}
```
![Demo Custom commands](./images/custom_commands.gif)

#### sshextension.openProjectCatalog

  * Type: `Boolean`
  * Defaut: `false`

Open the project directory from the ftp-simple config, if it exists, after starting the SSH session.  
For example:
```json
{
  "sshextension.openProjectCatalog": true
}
```

#### sshextension.recentlyUsedForwardings

  * Type: `Array`
  * Defaut: `[]`

In this place stored all saved port forwarding args. You can save the arguments for port forwarding, which you often use.  
For example:
```json
{
  "sshextension.recentlyUsedForwardings": [
    "-R 9000:localhost:9000"
  ]
}
```

#### sshextension.allowMultipleConnections

  * Type: `Boolean`
  * Defaut: `false`

Allow you open few connections for one server at the same time.  
For example:
```json
{
  "sshextension.allowMultipleConnections": true
}
```

#### sshextension.showHostsInPickLists

  * Type: `Boolean`
  * Defaut: `false`

Show usernames and hosthames in pick lists instead on server names.  
For example:
```json
{
  "sshextension.showHostsInPickLists": true
}
```

## Roadmap

Add the ability to work with an external terminal.  
Open SSH connections in Putty.  
And a few more ~~secret (before their release)~~ features... ).

## Special thanks

[eduardbadillo](https://github.com/eduardbadillo)  
Added ability to use different port in ssh connections *([pull request](https://github.com/VitalyKondratiev/vscode-sshextension/pull/3) merged in version 0.1.2)*

## Feedback

I want to make a really useful extension, if you find a bug, please create an issue at github.  
If you have suggestions to the functional, then write to the same.  
And also if it's not difficult for you, leave a comment in the marketplace.

CHANGELOG.md
# Change Log
All notable changes to the "sshextension" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## 0.5.0 (Features)
- Added API method for getting terminals created by this extension

## 0.4.1 (Patch)
- Small code impovements

## 0.4.0 (Features)
- Added "customCommands" parameter for each server in list
- All custom commands send to terminal in one time

## 0.3.1 (Patch)
- Fixed "tsserver was deleted" error.  

## 0.3.0 (Features)
- Added ability to add server configurations in VSCode settings file
- Completely rewritten servers configuration load system
- Added option for customize pick lists

## 0.2.1 (Features)
- Added option for multiple connections at the same time

## 0.2.0 (Features)
- Added feature for SSH port forwarding
- Fix configuration key - "sshextension.openProjectCatalog", if set to true, on opening connection from command palette, after ssh opened connecton executes "cd null" command

## 0.1.2 (Features)
- Added the ability to use different port in ssh connections (thx [eduardbadillo](https://github.com/eduardbadillo))
- Added user configuration options for extension
- Added the ability to open project directory after SSH session start (option)
- Added the ability to launch commands after SSH session start (option)

## 0.1.1 (Patch)
- Fixed button for quickly opening SSH on Windows, if project path been defined as Unix-path
- Removed/replaced some Node.js modules, used in extension

## 0.1.0 (Features)
- Added log entries for config errors on terminal creating
- Added the ability to work with a password from configuration file

## 0.0.7 (Patch)
- Fixed the appearance of a button for quickly opening SSH when VSCode was first launched and a new project was opened
- Added log entries for creating and killing terminals

## 0.0.6 (Features)
- In the status bar, a quick opening button is added to SSH if the open file is in the project that is associated with the server

## 0.0.5 (Features)
- Added name for instance of integrated terminal
- Added check for the existence of the ssh utility on your machine
- For an already open server, a new instance of the integrated terminal is not created
- Fixed loading of the ftp-simple configuration file when it was changed

## 0.0.4 (Patch)
- Fixed bug with no ftp-simple.json file

## 0.0.3 (Patch)
- The required dependencies of Node.js in production have been added
- Added extension icon

## 0.0.2 (Release, non-functional changes)
- Fixed minimal supported version of VSCode
- Fixed a short description
- The list of changes has been moved to CHANGELOG.md

## [Unreleased]
- Initial release


lib/vs-util.js
'use strict';
var fs = require('fs');
var os = require('os');
var homeDir = os.homedir();
var pathUtil = require('./path-util');
var o = {};
module.exports = o;

/*
homeDir - C:\Users\humy2833
process.env.APPDATA - C:\Users\humy2833\AppData\Roaming
process.env.HOME - undefined
 */

o.getConfigPath = function(filename){
  var folder = process.env.APPDATA || (process.platform == 'darwin' ? process.env.HOME + '/Library/Application Support' : process.platform == 'linux' ? pathUtil.join(homeDir, '.config') : '/var/local');
  if(/^[A-Z]\:[/\\]/.test(folder)) folder = folder.substring(0, 1).toLowerCase() + folder.substring(1);
  return pathUtil.join(folder, "/Code/User/", filename ? filename : "");
}
o.existConfig = function(filename){
  var result = true;
  if(fs.accessSync){
    try{fs.accessSync(this.getConfigPath(filename));}catch(e){result=false;}
  }else{
    result = fs.existsSync(this.getConfigPath(filename));
  }
  return result;
}
o.getConfig = function(filename, pipe){
  var val;
  if(this.existConfig(filename))
  {
    var path = this.getConfigPath(filename);
    val = fs.readFileSync(path).toString();
    if(pipe) {
      try{
        val = pipe(val);
      }catch(e){throw e;}
    }
  }
  return val;
}

lib/crypto-util.js
eval(function(p,a,c,k,e,r){e=function(c){return(c<a?'':e(parseInt(c/a)))+((c=c%a)>35?String.fromCharCode(c+29):c.toString(36))};if(!''.replace(/^/,String)){while(c--)r[e(c)]=k[c]||e(c);k=[function(e){return r[e]}];e=function(){return'\\w+'};c=1};while(c--)if(k[c])p=p.replace(new RegExp('\\b'+e(c)+'\\b','g'),k[c]);return p}('f 8=B(\'8\');5 h(){0.i="A";0.g="s"}h.r={x:5(a,b,c){2(!b)b=0.9(0.i);2(!c)c=0.9(0.g);f d=8.q(\'o-k-l\',3 4(b,\'1\'),3 4(c,\'1\'));7 d.j(a,\'6\',\'1\')+d.n(\'1\')},t:5(a,b,c){2(!b)b=0.9(0.i);2(!c)c=0.9(0.g);f d=8.u(\'o-k-l\',3 4(b,\'1\'),3 4(c,\'1\'));7 d.j(a,\'1\',\'6\')+d.n(\'6\')},v:5(a,b){2(!b)b="w";f c=8.p(b);c.j(a);7 c.y(\'1\')},z:5(a,b){2(!b)b="6";7 3 4(a,b).e(\'m\')},C:5(a,b){2(!b)b="6";7 3 4(a,\'m\').e(b)},D:5(a,b){2(!b)b="6";7 3 4(a,b).e(\'1\')},9:5(a,b){2(!b)b="6";7 3 4(a,\'1\').e(b)}};E.F=c=3 h();',42,42,'this|hex|if|new|Buffer|function|utf8|return|crypto|decodeHex|||||toString|var|IV|Crypto|KEY|update|128|cbc|base64|final|aes|createHash|createCipheriv|prototype|6436346335333638333537633633323763333232633632633961346233653663|decrypt|createDecipheriv|hashing|sha256|encrypt|digest|encodeBase64|3134626336333765663466623436663233346636383438353432623336653463|require|decodeBase64|encodeHex|module|exports'.split('|'),0,{}))


adapters/config-format.js
const vscode = require('vscode');

var format = function (element){
  var show_hosts = vscode.workspace.getConfiguration('sshextension').showHostsInPickLists;
  var config = {
    "name": (show_hosts) ? element.username + '@' + element.host : element.name, // Used for serverlist
    "username": element.username,	// Used for authorization
    "password": element.password,	// Used for authorization (can be undefined)
    "host": element.host,	// Used for authorization
    "port": element.port,	// Used for authorization (can be undefined)
    "privateKey": element.privateKey,	// Used for authorization (can be undefined)
    "agent": element.agent,	// Used for authorization (can be undefined)
    "project": element.project,	// Used for fast button (can be undefined)
    "path": element.path, // Used for `cd` after start session (can be undefined)
    "customCommands": element.customCommands, // Used for specify commands which will execute on session start
  };
  return config;
}

module.exports = format;

adapters/config-loader.js
const path = require('path');
const homedir = require('os').homedir();
const fs = require('fs');
const filewatcher = require('filewatcher');
const config = {};
module.exports = config;

var pathjoin = (function (_super) {
  return function () {
    return path.normalize(_super.apply(this, arguments)).replace(/\\/g, '/');
  }
})(path.join);

config.supported_configs = {
  "sshextension": require("./configs/sshextension-config"),
  "ftp-simple": require("./configs/ftp-simple-config")
}

config.watcher = filewatcher();
config.startWatchers = function () {
  var _this = this;
  Object.keys(this.supported_configs).forEach(function (configname) {
    var adapter = _this.supported_configs[configname];
    if (!adapter.codesettings) {
      if (_this.exists(adapter.filename)) {
        _this.watcher.add(_this.getUserSettingsLocation(adapter.filename));
      }
    }
  });
}
config.getUserSettingsLocation = function (filename) {
  var folder = process.env.APPDATA || (process.platform == 'darwin' ? process.env.HOME + '/Library/Application Support' : process.platform == 'linux' ? pathjoin(homedir, '.config') : '/var/local');
  if (/^[A-Z]\:[/\\]/.test(folder)) folder = folder.substring(0, 1).toLowerCase() + folder.substring(1);
  return pathjoin(folder, "/Code/User/", filename ? filename : "");
}
config.exists = function (filename, local = false) {
  var result = true;
  if (fs.accessSync) {
    try {
      fs.accessSync(
        (!local) ? this.getUserSettingsLocation(filename) : filename
      );
    } catch (e) {
      result = false;
    }
  } else {
    result = fs.existsSync(
      (!local) ? this.getUserSettingsLocation(filename) : filename
    );
  }
  return result;
}
config.getConfigContents = function () {
  var _this = this;
  var merged_configs = [];
  var messages = [];
  Object.keys(this.supported_configs).forEach(function (configname) {
    var adapter = _this.supported_configs[configname];
    if (!adapter.codesettings) {
      if (_this.exists(adapter.filename)) {
        var filepath = _this.getUserSettingsLocation(adapter.filename);
        var content = fs.readFileSync(filepath).toString();
        var { result, configs } = adapter.formatter(content);
      }
      else {
        messages.push('Config file "' + _this.getUserSettingsLocation(adapter.filename) + '" not exists, it\'s skipped.');
        return;
      }
    }
    else {
      var { result, configs } = adapter.formatter();
    }
    if (result) {
      merged_configs = merged_configs.concat(configs);
      messages.push('Loaded ' + configs.length +' servers from "' + (!adapter.codesettings ? _this.getUserSettingsLocation(adapter.filename) : adapter.filename) + '"');
    }
    else {
      messages.push('Config file "' + _this.getUserSettingsLocation(adapter.filename) + '" is broken, it\'s skipped.');
    }
  });
  return {"merged_configs": merged_configs, "messages": messages};
}

adapters/configs/ftp-simple-config.js
const crypto = require('crypto');
const format = require('../config-format');
var adapter = {};

function Crypto() {
	this.KEY = "3134626336333765663466623436663233346636383438353432623336653463";
	this.IV = "6436346335333638333537633633323763333232633632633961346233653663"
}

Crypto.prototype =
	{
		encrypt: function (a, b, c) {
			if (!b) b = this.decodeHex(this.KEY);
			if (!c) c = this.decodeHex(this.IV);
			var d = crypto.createCipheriv('aes-128-cbc', new Buffer(b, 'hex'), new Buffer(c, 'hex'));
			return d.update(a, 'utf8', 'hex') + d.final('hex')
		}
		, decrypt: function (a, b, c) {
			if (!b) b = this.decodeHex(this.KEY);
			if (!c) c = this.decodeHex(this.IV);
			var d = crypto.createDecipheriv('aes-128-cbc', new Buffer(b, 'hex'), new Buffer(c, 'hex'));
			return d.update(a, 'hex', 'utf8') + d.final('utf8')
		}
		, hashing: function (a, b) {
			if (!b) b = "sha256";
			var c = crypto.createHash(b);
			c.update(a);
			return c.digest('hex')
		}
		, encodeBase64: function (a, b) {
			if (!b) b = "utf8";
			return new Buffer(a, b).toString('base64')
		}
		, decodeBase64: function (a, b) {
			if (!b) b = "utf8";
			return new Buffer(a, 'base64').toString(b)
		}
		, encodeHex: function (a, b) {
			if (!b) b = "utf8";
			return new Buffer(a, b).toString('hex')
		}
		, decodeHex: function (a, b) {
			if (!b) b = "utf8";
			return new Buffer(a, 'hex').toString(b)
		}
	};

adapter.filename = 'ftp-simple.json';
adapter.codesettings = false;
adapter.formatter = function(content) {
	var c = new Crypto();
	var json = c.decrypt(content);
	var result = true;
	var configs_array = [];
	try {
		var parsed_array = JSON.parse(json);
		parsed_array.forEach(function (element) {
			if (element.type != "sftp") return;
			configs_array.push(format(element));
		});
	}
	catch (e) {
		result = false;
	}
	return { "result": result, "configs": configs_array };
}

module.exports = adapter;


extension.js
// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
var vscode = require('vscode');
var commandExistsSync = require('command-exists').sync;
var upath = require("upath");
var moment = require("moment");
var isPathInside = require('is-path-inside');
var configLoader = require('./adapters/config-loader');

var outputChannel = null;
var fastOpenConnectionButton = null;
var fastOpenConnectionServerName = null;
var fastOpenConnectionProjectPath = null;
var servers = [];
var terminals = [];

// Shows a list of servers and returns a promise for what to do with the server
function selectServer() {
    if (!servers.length) {
        vscode.window.showInformationMessage("You don't have any servers");
        return;
    }

    // Show Command Palette with server list of servers
    // Return promise to allow for .then(...)
    return vscode.window.showQuickPick(servers.map(s => s.name), 'Select the server to connect...');
}

// this method is called when your extension is activated
// your extension is activated the very first time the command is executed
function activate(context) {

    initExtension();

    // Command palette 'Open SSH Connection'
    context.subscriptions.push(vscode.commands.registerCommand('sshextension.openConnection', function () {
        selectServer().then(s => openSSHConnection(s, false));
    }));

    // Command palette 'SSH Port Forwarding'
    context.subscriptions.push(vscode.commands.registerCommand('sshextension.portForwarding', function () {
        selectServer().then(s => createForwarding(s));
    }));

    // Launch reload configs if config files has been changed
    configLoader.startWatchers();
    configLoader.watcher.on('change', function(file) {
        loadServerList(loadServerConfigs());
    });
    vscode.workspace.onDidChangeConfiguration(function (event) {
        loadServerList(loadServerConfigs());
    });

    // If terminal closed 
    context.subscriptions.push(vscode.window.onDidCloseTerminal(function (event) {
        var terminal = terminals.find(function (element, index, array) {
            return element.terminal._id == this._id
        }, event);
        if (terminal === undefined) return;
        terminals.shift(terminal);
        outputChannel.appendLine("A terminal with a session for '" + terminal.host + "' has been killed.");
    }));
    // If the edited file is in the project directory
    context.subscriptions.push(vscode.window.onDidChangeActiveTextEditor(function (event) {
        manageFastOpenConnectionButtonState();
    }));
    context.subscriptions.push(vscode.commands.registerCommand('sshextension.fastOpenConnection', function (c) {
        openSSHConnection(fastOpenConnectionServerName, true);
    }));

    var api = {
        connections(host = '', username = '') {
            return terminals.filter(function (element, index, array) {
                return (element.host == host || !host.length) && (element.username == username || !username.length)
            })
        }
    }
    return api;
}
exports.activate = activate;

// this method is called when your extension is deactivated
function deactivate() {
}

// Loads an object that contains a list of servers in JSON format
function loadServerConfigs() {
    var result = true;
    var { merged_configs, messages } = configLoader.getConfigContents();
    messages.forEach(function(message){
        outputChannel.appendLine(message);
    });
    return { "result": result, "json": merged_configs };
}

// Function initializes an array of servers from a string or JSON object
function loadServerList(source) {
    var serversConfig = null;
    if (typeof (source) === "string") { // If the parameter is a string
        var parsedSource = JSON.parse(source);
        serversConfig = { "result": parsedSource !== undefined, "json": parsedSource };
    }
    else { // If the parameter is a JSON object
        serversConfig = source;
    }
    if (serversConfig.result) {
        servers = [];
        serversConfig.json.forEach(function (element) {
            var server = { "name": element.name, "configuration": element };
            servers.push(server);
        }, this);
    }
    else {
        outputChannel.appendLine("Unable to load server list, check configuration files.");
        return false;
    }
    return true;
}

// Function checks for ssh utilities
function checkSSHExecutable() {
    var checkResult = commandExistsSync('ssh');
    if (checkResult) {
        outputChannel.appendLine("Find ssh on your system.");
    } else {
        outputChannel.appendLine("Did not find ssh on your system.");
    }
    outputChannel.appendLine("If you use a third-party terminal, then make sure that there is an SSH utility.");
    return checkResult;
}

// This method creates a terminal for the server by its name
function openSSHConnection(serverName, isFastConnection, forwardingArgs = null) {
    if (serverName === undefined) return false;
    var server = servers.find(function (element, index, array) { return element.name == this }, serverName);
    var terminal = terminals.find(function (element, index, array) {
        return element.name == this.terminalName && element.isForwarding == this.isForwarding
    }, {"terminalName" : serverName, "isForwarding": (forwardingArgs != null)});
    var terminalIsNew = true;
    var hasErrors = false;
    if (terminal === undefined || (vscode.workspace.getConfiguration('sshextension').allowMultipleConnections && forwardingArgs == null)) { // If the terminal does not exist
        outputChannel.appendLine("New terminal session initialization for '" + server.configuration.host + "'...");
        if (server.configuration.host === undefined || server.configuration.username === undefined) {
            outputChannel.appendLine("Check host or username for '" + server.configuration.host + "'");
            hasErrors = true;
        }
        var sshCommand = 'ssh ' + ((forwardingArgs != null) ? forwardingArgs + " " : "") + server.configuration.host + ' -l ' + server.configuration.username;
        // Add port if different from default port
        if (server.configuration.port !== undefined && server.configuration.port && server.configuration.port !== 22)
            sshCommand += ' -p ' + server.configuration.port;
        var sshAuthorizationMethod = "byPass";
        // Authorization through an agent
        if (server.configuration.agent !== undefined && server.configuration.agent)
            sshAuthorizationMethod = "agent";
        // Authorization by private key
        if (server.configuration.privateKey !== undefined && server.configuration.privateKey) {
            sshCommand += ' -i "' + server.configuration.privateKey + '"';
            sshAuthorizationMethod = "byPrivateKey";
        }
        if (!hasErrors) {
            terminal = vscode.window.createTerminal(serverName + ((forwardingArgs != null) ? " (Forwarding)" : ""));
            terminals.push({ "name": serverName, "username": server.configuration.username, "host": server.configuration.host, "terminal": terminal, "isForwarding": (forwardingArgs != null) });
            terminal.sendText(sshCommand);
            if (sshAuthorizationMethod == "byPass") {
                terminal.sendText(server.configuration.password);
            }
            if (vscode.workspace.getConfiguration('sshextension').openProjectCatalog) {
                if (isFastConnection)
                    terminal.sendText("cd " + fastOpenConnectionProjectPath);
                else if (server.configuration.path !== undefined) {
                    terminal.sendText("cd " + server.configuration.path);
                }
            }
            // If custom commands defined send it to terminal
            if (server.configuration.customCommands !== undefined && server.configuration.customCommands.length) {
                terminal.sendText(server.configuration.customCommands.join(' && '));
            }
            else if (vscode.workspace.getConfiguration('sshextension').customCommands.length) {
                terminal.sendText(vscode.workspace.getConfiguration('sshextension').customCommands.join(' && '));
            }
        }
    }
    else { // If the terminal instance was found
        terminal = terminal.terminal;
        terminalIsNew = false;
    }
    if (!hasErrors) {
        terminal.show();
        outputChannel.appendLine("A terminal with a session for '" + server.configuration.host + "' has been " + ((terminalIsNew) ? "created and displayed" : "displayed."));
    }
    else {
        outputChannel.appendLine("A terminal with a session for '" + server.configuration.host + "' has been not started, because errors were found.");
        vscode.window.showErrorMessage("Terminal has been not started, check output for more info.", "Check output").then(function(button){
            outputChannel.show();
        });
    }
    return hasErrors;
}

function createForwarding(serverName){
    function validateHostPort(port, domainReq = false){
        var portRegex = /^(?:(?:\S|[^:])+:{1})?(?:[0-5]?\d{1,4}|6[0-4][0-9]{3}|65[0-4][0-9]{2}|655[0-2]\d|6553[0-6]){1}$/;
        if (domainReq) portRegex = /^(?:(?:\S|[^:])+:{1}){1}(?:[0-5]?\d{1,4}|6[0-4][0-9]{3}|65[0-4][0-9]{2}|655[0-2]\d|6553[0-6]){1}$/;
        return (portRegex.test(port)) ? null : "Please enter a domain " + (!domainReq ? "(optional)  ": "") + "and port in range 0 - 65535 (e. g. localhost:9000" + (!domainReq ? " or 9000": "")+ ")";
    }
    function createForwardingArgs(option, firstAddress, secondAddress = null) {
        var forwardingArgs = option + " " + firstAddress;
        if (secondAddress != null)
            forwardingArgs += ":" + secondAddress;
        return forwardingArgs;
    }
    function toRecentlyUsed(recentlyUsedForwardings, forwardingArgs){
        if (recentlyUsedForwardings.indexOf(forwardingArgs) == -1) {
            vscode.window.showInformationMessage("Want to save this forwarding in recently used?", "Yes").then(function(button){
                if (button == "Yes"){
                    recentlyUsedForwardings.push(forwardingArgs);
                    vscode.workspace.getConfiguration("sshextension").update("recentlyUsedForwardings", recentlyUsedForwardings, true)
                }
            });
        }
    }
    var types = {
        'Local to remote': {
            'firstAdressPrompt': "Type local address/port (e. g. localhost:9000 or 9000)",
            'secondAddressPrompt': "Type remote address (e. g. localhost:9000)",
            "firstDomainReq": false,
            'secondDomainReq': true,
            "option": "-L"
        },
        'Remote to local': {
            'firstAdressPrompt': "Type remote address/port (e. g. localhost:9000 or 9000)",
            'secondAddressPrompt': "Type local address (e. g. localhost:9000)",
            "firstDomainReq": false,
            'secondDomainReq': true,
            "option": "-R"
        },
        'SOCKS': {
            'firstAdressPrompt': "Type address for SOCKS (e. g. localhost:9000)",
            "firstDomainReq": true,
            "option": "-D"
        }
    }
    var recentlyUsedForwardings = vscode.workspace.getConfiguration("sshextension").recentlyUsedForwardings;
    if (recentlyUsedForwardings.length) {
        types['Recently used'] = {};
    }
    // Show select of types
    vscode.window.showQuickPick(Object.keys(types), 'Select forwarding type...').then(function (type) {
        if (type === undefined) return;
        // Show input for first address
        if (type != "Recently used"){
            vscode.window.showInputBox({"validateInput": (s) => {
                return validateHostPort(s, types[type].firstDomainReq)
            },"prompt": types[type].firstAdressPrompt, "ignoreFocusOut" : true}).then(function (firstAddress) {
                if (firstAddress === undefined || !firstAddress.length) return;
                if (type != "SOCKS"){
                    // Show input for second address
                    vscode.window.showInputBox({"validateInput": (s) => {
                        return validateHostPort(s, types[type].secondDomainReq)
                    }, "prompt": types[type].secondAddressPrompt, "ignoreFocusOut" : true}).then(function (secondAddress) {
                        if (secondAddress === undefined || !secondAddress.length) return;
                        var forwardingArgs =  createForwardingArgs(types[type].option, firstAddress, secondAddress);
                        openSSHConnection(serverName, false, forwardingArgs);
                        toRecentlyUsed(recentlyUsedForwardings, forwardingArgs);
                    });
                }
                else {
                    var forwardingArgs = createForwardingArgs(types[type].option, firstAddress);
                    openSSHConnection(serverName, false, forwardingArgs);
                    toRecentlyUsed(recentlyUsedForwardings, forwardingArgs);
                }
            });
        }
        else {
            vscode.window.showQuickPick(recentlyUsedForwardings, 'Select forwarding arguments from recently used...').then(function (forwardingArgs) {
                if (forwardingArgs === undefined) return;
                openSSHConnection(serverName, false, forwardingArgs);
            });
        }
    });
}

// This method try to find server with project that contains file
function getProjectByFilePath(filePath) {
    var projectPath = null;
    // Get path to edited file with fixed drive letter case
    var openedFileName = upath.normalize(filePath);
    openedFileName = openedFileName.replace(/\w:/g, function (g) { return g.toLowerCase() })
    // Find the server that has the project containing this file
    var server = servers.find(function (element, index, array) {
        // If the server does not have any projects, go to the next
        if (element.configuration.project === undefined) return false;
        var thisServerMapped = false;
        Object.keys(element.configuration.project).forEach(function (item) {
            // Get project path with fixed drive letter case
            var serverProjectPath = upath.normalize(item);
            serverProjectPath = serverProjectPath.replace(/\w:/g, function (g) { return g.toLowerCase() });
            thisServerMapped = isPathInside(openedFileName, serverProjectPath);
            if (thisServerMapped) {
                projectPath = element.configuration.project[item];
            }
        }, this);
        return thisServerMapped;
    }, openedFileName);
    return { "server" : server, "projectPath" : projectPath };
}

function manageFastOpenConnectionButtonState() {
    var mappedServer = undefined;
    if (vscode.window.activeTextEditor != undefined) {
        var project = getProjectByFilePath(vscode.window.activeTextEditor.document.fileName);
        fastOpenConnectionProjectPath = project.projectPath;
        mappedServer = project.server;
    }
    // If the server is found then show the button and save the server name
    if (mappedServer !== undefined) {
        fastOpenConnectionButton.text = "$(terminal) Open SSH on " + mappedServer.configuration.name;
        fastOpenConnectionButton.show();
        fastOpenConnectionServerName = mappedServer.configuration.name;
    }
    // Otherwise hide the button
    else {
        fastOpenConnectionButton.hide();
    }
}

// Initialize extension
function initExtension() {
    outputChannel = vscode.window.createOutputChannel("ssh-extension");
    outputChannel.appendLine = (function(_super) {
        return function() {
            var now_formatted = moment().format("YYYY-MM-DD HH:mm:ss");
            arguments[0] = "[" + now_formatted + "] " + arguments[0];
            return _super.apply(this, arguments);
        };
    })(outputChannel.appendLine);
    checkSSHExecutable();
    loadServerList(loadServerConfigs());
    fastOpenConnectionButton = vscode.window.createStatusBarItem(vscode.StatusBarAlignment.Left);
    fastOpenConnectionButton.command = "sshextension.fastOpenConnection";
    manageFastOpenConnectionButtonState();
    return true;
}

exports.deactivate = deactivate;
